core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -273,35 +273,28 @@ private[history] class FsHistoryProvider(conf: SparkConf) extends ApplicationHis
    */
   private def cleanLogs(): Unit = {
     try {
-      val statusList = Option(fs.listStatus(new Path(logDir))).map(_.toSeq)
-        .getOrElse(Seq[FileStatus]())
       val maxAge = conf.getLong("spark.history.fs.cleaner.maxAge.seconds",
         DEFAULT_SPARK_HISTORY_FS_MAXAGE_S) * 1000
 
       val now = System.currentTimeMillis()
       val appsToRetain = new mutable.LinkedHashMap[String, FsApplicationHistoryInfo]()
 
+      // Scan all logs from the log directory.
+      // Only completed applications older than the specified max age will be deleted.
       applications.values.foreach { info =>
-        if (now - info.lastUpdated <= maxAge) {
+        if (now - info.lastUpdated <= maxAge || !info.completed) {
           appsToRetain += (info.id -> info)
+        } else {
+          try {
+            fs.delete(new Path(logDir + "/" + info.logPath), true)
+          } catch {
+            case t: IOException => logError(s"IOException in cleaning logs of ${info.logPath}", t)
+            appsToRetain += (info.id -> info)
```

Comment:
Shouldn't this line be indented? And the `logError` above me on its own line, at the same indent level as this one?

If feels a little weird to leave the app on the live list; in the case of old-style logs, a few files may have been deleted, while others remain, so the app would be "un-renderable". On the other hand, this means the code will try to clean up this app again later on, so I guess it's better this way.